### DISCUSSION TITLE
메인하트버튼에 햅틱을 주게 된 개발자이야기

### DISCUSSION THREAD

**Discussion by chaneeii** [Team-WeHere/WeHere#5] [Ideas]
## Intro

테오가 다음과 같은 커멘트를 주었는데, 
- https://github.com/Team-WeHere/WeHere/pull/3#discussion_r1014710520

제가 너무 심심해서 잔재주를 부린거라 `반영이 안되어도 이러면 쫌 재밌네` 하고 추가한 햅틱피드백입니다 
저는 다음과 같은 생각들을하다가 추가를 해보게 되었습니다. (다소 개발적 이야기도 담긴)

## 생각의 흐름

**1. 버튼을 깜빡였을때 오류로 보이는, 불안정한 느낌? 이 든다!**

1-1. 기술적인 한계
사실 아시겠지만 스유나, uikit 의 기본 버튼은 다 테오가 말씀해주신것처럼 깜빡이는 애니메이션을 적용하는데요!

가운데 하트버튼은 기본버튼으로 클릭시 깜박이도록 처리하고
나머지 작은 주위에 둘러싼 버튼들은 opacity를 0에서 100으로 버튼 클릭에 따라 자연스럽게 변화하도록 설정했어요

애니메이션은 언제까지나 눈속임이기에 간단히 원리만 설명하자면 하트뒤에 작은 버튼들이 버튼들은 **한곳에 겹쳐 쌓여 숨겨져** 있다가
버튼을 누르면 각자의 각도와 지정된 위치로 하나의 속도로 이동합니다. 

<img width="632" alt="스크린샷 2022-11-06 오후 1 21 47" src="https://user-images.githubusercontent.com/63157395/200154033-3e7b8125-3723-4d0b-98fb-cec7a4180e99.png">

1-2. 시각적으로 불안정함
fab 는 항상 동동떠다니며 디자인적으로 항상 고정된 위치에 있어야하는 요소인데
깜박거리니 뷰에서 너무 사라지는 것 같아서 불안정한 느낌이 들었어요.

**🥳 사용자가 기대대로 터치를 하는 경우**

https://user-images.githubusercontent.com/63157395/200153359-fdf60e7d-4c00-4073-bcf3-53c0b5492dc1.MP4

**😫 사용자가 터치를 기대와 다르게 꾸우욱 계속 누르는 코드**

하트뒤에 작은 버튼들이 버튼들은 **한곳에 겹쳐 쌓여 숨어** 있던 버튼들이 노출이 되기도 합니다!

https://user-images.githubusercontent.com/63157395/200153362-3ecd94e6-28dd-4552-806e-71339fd764c0.MP4

**2. 기존 앱들의 Fab 버튼 확장에서 어떠한 식으로 대응하는가...?**

하지만 이분들은 저와 다르게 실력이 출중하시고, 아마 스유대신 다른걸 건드려서 깜박이는 처리를 하더라도 뒤에 애기 버튼들이 안보이게 하실수있었겠지만, 나는 모르는 이 버튼들을 계속 띄워둔 이유가 있지 않을까 싶어 이런저런 레퍼들을 찾았는데

제일먼저 네이버의 그린닷을 보면 버튼이 바로 `터치 순간에 반응`해서 `촥! 펼쳐지`고 `햅틱`이 들어오는데 약간 쾌감 때문에 재밌어서 버튼을 누르게 되더라구요!
https://youtu.be/B8niucICdTs?t=22

그래서 찾아보니 Fab 버튼에 햅틱, 진동반응을 넣는 것들을 찾아볼수있었어요
- 오페라 : https://help.opera.com/en/touch/fast-action-button/
<img src="https://user-images.githubusercontent.com/63157395/200153775-195fe750-adfc-4b41-a8a0-af540233c801.png" width="1000" >

- 플러터 : https://api.flutter.dev/flutter/material/FloatingActionButton-class.html
<img src="https://user-images.githubusercontent.com/63157395/200153797-bc5e35b2-7103-4ddd-9f3e-de99f3b7d4f9.png" width="1000" >

3. 오 햅틱을 넣어보자!
그래서 햅틱을 넣어보게 되었습니다

## 어떻게 햅틱을 넣었는가?

저희 버튼들은 자연스럽게 튀어나오기 위해서 약간의 딜레이를 주는데요
버튼을 누르자마자 나오면 약간 자연스럽지 못함 + 경박함이 느껴져서 약간 딜레이를 주었는데

그러다 보니 버튼을 누르자 마자 햅틱을 주는게 아닌 펼쳐지는 순간에 작은 버튼들에 맞춰 햅틱을 주게 되었습니다.

그리고 hig에서 햅틱 그래프를 보고 버튼이 펼쳐지는 것과 유사한 형태인 success 햅틱을 주게 되었어요 (커지는 효과)

![스크린샷 2022-11-06 오후 1 29 01](https://user-images.githubusercontent.com/63157395/200154228-a643cc0c-32fa-4e7e-87ad-d30edd7a9cc6.png)

- https://developer.apple.com/design/human-interface-guidelines/patterns/playing-haptics/

## 끝!

여기까지 햅틱을 어떻게 주게 되었는지 생각을 정리해보았습니다!
요거 클래스를 정의해놔서 아무때나 재사용이 가능하니 햅틱이 필요하다면 필요한 순간을 적절히 잘 설계해주시면 바로 반영가능이요 ✨

**Comment by phainestha1** [comment]
긴 시간 고민하신 흔적이 느껴집니다!
앞으로 업무할 때 이정도로 깊게 생각하고 집중할 수 있도록 해야겠어요 🥺
감사합니다!
